### ISSUE TITLE
Unexpected error

### ISSUE THREAD

**Issue by SpookyBoogieLondon** [microsoft/pyright#2890]
**Describe the bug**
The code provided procudes an error (in basic mode), which seems incorrect to me.

The error is:

```
Argument of type "bool" cannot be assigned to parameter "b" of type "_VT@dict" in function "__init__"
  Type "list[Unknown]" cannot be assigned to type "bool"
```

**To Reproduce**
Use code provided and pyright ver `1.1.211`
The aforementioned message appears from version `1.1.184` onwards. Before that version, the message was:
```
error: No overloads for "__init__" match the provided arguments
```

**Expected behavior**
No errors.

**Code**

```python
def foo(u):
    b: bool = True

    return dict(
        b=b,
        u=u,  #  type annotate 'u' in params, remove this line or...
        x=[], #  remove this line to fix
        )
```

**Comment by erictraut** [comment]
Thanks for the bug report. The TypeVar constraint solver in pyright contains various heuristics the deal with unknown (unannotated) values, but one of these heuristics was failing to handle the case in your example. I've tweaked the heuristic so it now works. This will be addressed in the next release.

In the meantime, you've already identified several workarounds.

**Comment by SpookyBoogieLondon** [comment]
Great stuff. Thank you for being so prompt! 👍

**Comment by erictraut** [comment]
Leaving the issue open until the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.212, which I just published. It will also be included in the next release of pylance.

**Comment by zevaverbach** [comment]
Hey, I'm afraid this doesn't seem to be fixed. I ran into the same error using 1.1.220 in Neovim 0.6.1 just now.

The dictionary looks like this:

```
data = dict(invoice=dict(
        customerid=client_id, 
        notes=notes, 
        po_number=po_number,
        status=status,
        lines=lines,
        create_date=create_date,
        allowed_gateway_name="Stripe",
    ))
```

PyRight is grumpy about `client_id`, `status`, and `lines`, which are defined like so in the containing function's arguments:

```
def func(client_id: int, notes: str, po_number: str, lines: List[dict], status: Union[str, int], create_date=None):
    ...
```

So maybe this has reverted?
